In [ ]:
import pandas as pd
import datetime
import numpy as np
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)

In [ ]:
contacts_df = pd.read_excel('Contacts.xlsx', dtype={'Id': str})
calls_df = pd.read_excel('Calls.xlsx', dtype={'Id': str, "CONTACTID": str})
spend_df = pd.read_excel('Spend.xlsx')
deals_df = pd.read_excel('Deals.xlsx', dtype={'Id': str, 'Contact Name': str})

***Очистка и подготовка данных:***

# **Contacts**


In [ ]:
print("Первые 5 строк:")
display(contacts_df.head())

print("Информация о датасете:")
display(contacts_df.info())

print("Описательная статистика:")
display(contacts_df.describe())

print("Количество пропущенных значений в каждом столбце:")
display(contacts_df.isnull().sum())

Первые 5 строк:


Id Contact Owner Name      Created Time     Modified Time
0  5805028000000645014       Rachel White  27.06.2023 11:28  22.12.2023 13:34
1  5805028000000872003      Charlie Davis  03.07.2023 11:31  21.05.2024 10:23
2  5805028000000889001          Bob Brown  02.07.2023 22:37  21.12.2023 13:17
3  5805028000000907006          Bob Brown  03.07.2023 05:44  29.12.2023 15:20
4  5805028000000939010         Nina Scott  04.07.2023 10:11  16.04.2024 16:14

Информация о датасете:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18548 entries, 0 to 18547
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Id                  18548 non-null  object
 1   Contact Owner Name  18548 non-null  object
 2   Created Time        18548 non-null  object
 3   Modified Time       18548 non-null  object
dtypes: object(4)
memory usage: 579.8+ KB


None

Описательная статистика:


Id Contact Owner Name      Created Time  \
count                 18548              18548             18548   
unique                18548                 28             17921   
top     5805028000000645014      Charlie Davis  10.06.2024 09:00   
freq                      1               2018                13   

           Modified Time  
count              18548  
unique             16580  
top     13.06.2024 17:08  
freq                  25

Количество пропущенных значений в каждом столбце:


Id                    0
Contact Owner Name    0
Created Time          0
Modified Time         0
dtype: int64

1. Удалите дубликаты и неактуальные столбцы.


In [ ]:
contacts_df[
    contacts_df.duplicated(subset=contacts_df.columns[1:])
].shape

(38, 4)

In [ ]:
contacts_df.drop_duplicates(
    subset=contacts_df.columns[1:],
    inplace=True
)
contacts_df[
    contacts_df.duplicated(subset=contacts_df.columns[1:])
].shape

(0, 4)

2. Соответствующим образом обработайте отсутствующие значения


In [ ]:
unique_values = (
    contacts_df["Contact Owner Name"]
    .unique()
    .tolist()
)
unique_values

['Rachel White',
 'Charlie Davis',
 'Bob Brown',
 'Nina Scott',
 'Alice Johnson',
 'Ian Miller',
 'Jane Smith',
 'Julia Nelson',
 'George King',
 'Quincy Vincent',
 'Diana Evans',
 'Kevin Parker',
 'Ulysses Adams',
 'Victor Barnes',
 'Yara Edwards',
 'Paula Underwood',
 'Mason Roberts',
 'Ben Hall',
 'Amy Green',
 'Cara Iverson',
 'Oliver Taylor',
 'Eva Kent',
 False,
 'Zachary Foster',
 'Sam Young',
 'Wendy Clark',
 'Tina Zhang',
 'Derek James']

*Строки, где значение 'Contact Owner Name' равно False,  может указывать на некорректные записи. Удаляем их.*

In [ ]:
contacts_df = contacts_df[
    contacts_df['Contact Owner Name'] != False
]

3. Преобразуйте типы данных для таких столбцов, как даты и числовые значения.

In [ ]:
contacts_df["Created Time"] = pd.to_datetime(
    contacts_df["Created Time"],
    errors="raise"
)
contacts_df["Modified Time"] = pd.to_datetime(
    contacts_df["Modified Time"],
    errors="raise"
)

In [ ]:
display(contacts_df.info(), contacts_df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
Index: 18509 entries, 0 to 18547
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Id                  18509 non-null  object        
 1   Contact Owner Name  18509 non-null  object        
 2   Created Time        18509 non-null  datetime64[ns]
 3   Modified Time       18509 non-null  datetime64[ns]
dtypes: datetime64[ns](2), object(2)
memory usage: 723.0+ KB


None

Id                    0
Contact Owner Name    0
Created Time          0
Modified Time         0
dtype: int64

In [ ]:
contacts_df.to_excel(
    '/content/Processed_Contacts_df.xlsx',
    index=False
)

# **Calls_df**




In [ ]:
print("Первые 5 строк:")
display(calls_df.head())

print("Информация о датасете:")
display(calls_df.info())

print("Описательная статистика:")
display(calls_df.describe())

print("Количество пропущенных значений в каждом столбце:")
print(calls_df.isnull().sum())

Первые 5 строк:


Id   Call Start Time Call Owner Name            CONTACTID  \
0  5805028000000805001  30.06.2023 08:43        John Doe                  NaN   
1  5805028000000768006  30.06.2023 08:46        John Doe                  NaN   
2  5805028000000764027  30.06.2023 08:59        John Doe                  NaN   
3  5805028000000787003  30.06.2023 09:20        John Doe  5805028000000645014   
4  5805028000000768019  30.06.2023 09:30        John Doe  5805028000000645014   

  Call Type  Call Duration (in seconds)       Call Status  Dialled Number  \
0   Inbound                       171.0          Received             NaN   
1  Outbound                        28.0  Attended Dialled             NaN   
2  Outbound                        24.0  Attended Dialled             NaN   
3  Outbound                         6.0  Attended Dialled             NaN   
4  Outbound                        11.0  Attended Dialled             NaN   

  Outgoing Call Status  Scheduled in CRM  Tag  
0                  NaN               NaN  NaN  
1            Completed               0.0  NaN  
2            Completed               0.0  NaN  
3            Completed               0.0  NaN  
4            Completed               0.0  NaN

Информация о датасете:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95874 entries, 0 to 95873
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Id                          95874 non-null  object 
 1   Call Start Time             95874 non-null  object 
 2   Call Owner Name             95874 non-null  object 
 3   CONTACTID                   91941 non-null  object 
 4   Call Type                   95874 non-null  object 
 5   Call Duration (in seconds)  95791 non-null  float64
 6   Call Status                 95874 non-null  object 
 7   Dialled Number              0 non-null      float64
 8   Outgoing Call Status        86875 non-null  object 
 9   Scheduled in CRM            86875 non-null  float64
 10  Tag                         0 non-null      float64
dtypes: float64(4), object(7)
memory usage: 8.0+ MB


None

Описательная статистика:


Call Duration (in seconds)  Dialled Number  Scheduled in CRM  Tag
count                95791.000000             0.0      86875.000000  0.0
mean                   164.977263             NaN          0.001635  NaN
std                    401.410826             NaN          0.040397  NaN
min                      0.000000             NaN          0.000000  NaN
25%                      4.000000             NaN          0.000000  NaN
50%                      8.000000             NaN          0.000000  NaN
75%                     98.000000             NaN          0.000000  NaN
max                   7625.000000             NaN          1.000000  NaN

Количество пропущенных значений в каждом столбце:
Id                                0
Call Start Time                   0
Call Owner Name                   0
CONTACTID                      3933
Call Type                         0
Call Duration (in seconds)       83
Call Status                       0
Dialled Number                95874
Outgoing Call Status           8999
Scheduled in CRM               8999
Tag                           95874
dtype: int64


1. Удалите дубликаты и неактуальные столбцы.

In [ ]:
calls_df[
    calls_df.duplicated(subset=calls_df.columns[1:])
].shape

(3257, 11)

In [ ]:
calls_df.loc[
    calls_df.duplicated(subset=calls_df.columns[1:]),
    :
]

Id   Call Start Time Call Owner Name  \
35     5805028000001167001  06.07.2023 17:15   Alice Johnson   
102    5805028000001348077  08.07.2023 16:43        John Doe   
255    5805028000001546030  12.07.2023 19:23      Jane Smith   
278    5805028000001679013  13.07.2023 12:26      Jane Smith   
348    5805028000001792001  13.07.2023 20:28   Alice Johnson   
...                    ...               ...             ...   
95522  5805028000056725508  20.06.2024 17:43    Kevin Parker   
95692  5805028000056828261  21.06.2024 12:11      Ian Miller   
95804  5805028000056832311  21.06.2024 14:17    Yara Edwards   
95834  5805028000056873560  21.06.2024 14:47   Ulysses Adams   
95839  5805028000056892435  21.06.2024 14:55        John Doe   

                 CONTACTID Call Type  Call Duration (in seconds)  \
35     5805028000001129001  Outbound                         0.0   
102                    NaN    Missed                         0.0   
255    5805028000001552025  Outbound                         0.0   
278                    NaN    Missed                         0.0   
348    5805028000001363222  Outbound                         0.0   
...                    ...       ...                         ...   
95522  5805028000037097355    Missed                         0.0   
95692  5805028000056578096  Outbound                         0.0   
95804                  NaN  Outbound                         8.0   
95834  5805028000026041053  Outbound                         0.0   
95839                  NaN    Missed                         0.0   

              Call Status  Dialled Number Outgoing Call Status  \
35     Unattended Dialled             NaN            Completed   
102                Missed             NaN                  NaN   
255    Unattended Dialled             NaN            Completed   
278                Missed             NaN                  NaN   
348    Unattended Dialled             NaN            Completed   
...                   ...             ...                  ...   
95522              Missed             NaN                  NaN   
95692  Unattended Dialled             NaN            Completed   
95804    Attended Dialled             NaN            Completed   
95834  Unattended Dialled             NaN            Completed   
95839              Missed             NaN                  NaN   

       Scheduled in CRM  Tag  
35                  0.0  NaN  
102                 NaN  NaN  
255                 0.0  NaN  
278                 NaN  NaN  
348                 0.0  NaN  
...                 ...  ...  
95522               NaN  NaN  
95692               0.0  NaN  
95804               0.0  NaN  
95834               0.0  NaN  
95839               NaN  NaN  

[3257 rows x 11 columns]

In [ ]:
calls_df.drop_duplicates(
    subset=calls_df.columns[1:],
    inplace=True
)
calls_df[
    calls_df.duplicated(subset=calls_df.columns[1:])
].shape

(0, 11)

In [ ]:
calls_df.drop(
    columns=['Dialled Number', 'Tag'],
    inplace=True,
    errors='raise'
)

3. Преобразуйте типы данных для таких столбцов, как даты и числовые значения.


In [ ]:
calls_df['Call Start Time'] = pd.to_datetime(
    calls_df['Call Start Time'],
    errors='raise'
)

In [ ]:
display(calls_df.info(), calls_df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
Index: 92617 entries, 0 to 95873
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Id                          92617 non-null  object        
 1   Call Start Time             92617 non-null  datetime64[ns]
 2   Call Owner Name             92617 non-null  object        
 3   CONTACTID                   88815 non-null  object        
 4   Call Type                   92617 non-null  object        
 5   Call Duration (in seconds)  92538 non-null  float64       
 6   Call Status                 92617 non-null  object        
 7   Outgoing Call Status        83804 non-null  object        
 8   Scheduled in CRM            83804 non-null  float64       
dtypes: datetime64[ns](1), float64(2), object(6)
memory usage: 7.1+ MB


None

Id                               0
Call Start Time                  0
Call Owner Name                  0
CONTACTID                     3802
Call Type                        0
Call Duration (in seconds)      79
Call Status                      0
Outgoing Call Status          8813
Scheduled in CRM              8813
dtype: int64

In [ ]:
calls_df.to_excel(
    '/content/Processed_Calls_df.xlsx',
    index=False
)

# **Deals_df**

In [ ]:
print("Первые 5 строк:")
display(deals_df.head())

print("Информация о датасете:")
display(deals_df.info())

print("Описательная статистика:")
display(deals_df.describe())

print("Количество пропущенных значений в каждом столбце:")
print(deals_df.isnull().sum())

Первые 5 строк:


Id Deal Owner Name Closing Date            Quality  \
0  5805028000056864695        Ben Hall          NaN                NaN   
1  5805028000056859489   Ulysses Adams          NaN                NaN   
2  5805028000056832357   Ulysses Adams   21.06.2024     D - Non Target   
3  5805028000056824246        Eva Kent   21.06.2024  E - Non Qualified   
4  5805028000056873292        Ben Hall   21.06.2024     D - Non Target   

      Stage     Lost Reason       Page                  Campaign       SLA  \
0  New Lead             NaN  /eng/test             03.07.23women       NaN   
1  New Lead             NaN    /at-eng                       NaN       NaN   
2      Lost      Non target    /at-eng                engwien_AT  00:26:43   
3      Lost  Invalid number       /eng  04.07.23recentlymoved_DE  01:00:04   
4      Lost      Non target       /eng              discovery_DE  00:53:12   

              Content           Term          Source Payment Type  \
0                 v16          women    Facebook Ads          NaN   
1                 NaN            NaN         Organic          NaN   
2               b1-at     21_06_2024  Telegram posts          NaN   
3  bloggersvideo14com  recentlymoved    Facebook Ads          NaN   
4             website            NaN      Google Ads          NaN   

         Product Education Type      Created Time  Course duration  \
0            NaN            NaN  21.06.2024 15:30              NaN   
1  Web Developer        Morning  21.06.2024 15:23              6.0   
2            NaN            NaN  21.06.2024 14:45              NaN   
3            NaN            NaN  21.06.2024 13:32              NaN   
4            NaN            NaN  21.06.2024 13:21              NaN   

   Months of study Initial Amount Paid Offer Total Amount  \
0              NaN                 NaN                NaN   
1              NaN                   0               2000   
2              NaN                 NaN                NaN   
3              NaN                 NaN                NaN   
4              NaN                 NaN                NaN   

          Contact Name City Level of Deutsch  
0  5805028000056849495  NaN              NaN  
1  5805028000056834471  NaN              NaN  
2  5805028000056854421  NaN              NaN  
3  5805028000056889351  NaN              NaN  
4  5805028000056876176  NaN              NaN

Информация о датасете:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21595 entries, 0 to 21594
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   21593 non-null  object 
 1   Deal Owner Name      21564 non-null  object 
 2   Closing Date         14645 non-null  object 
 3   Quality              19340 non-null  object 
 4   Stage                21593 non-null  object 
 5   Lost Reason          16124 non-null  object 
 6   Page                 21593 non-null  object 
 7   Campaign             16067 non-null  object 
 8   SLA                  15533 non-null  object 
 9   Content              14147 non-null  object 
 10  Term                 12454 non-null  object 
 11  Source               21593 non-null  object 
 12  Payment Type         496 non-null    object 
 13  Product              3592 non-null   object 
 14  Education Type       3300 non-null   object 
 15  Created Time 

None

Описательная статистика:


Course duration  Months of study
count      3587.000000       840.000000
mean         10.198495         5.442857
std           1.834681         2.919920
min           6.000000         0.000000
25%          11.000000         3.000000
50%          11.000000         5.000000
75%          11.000000         8.000000
max          11.000000        11.000000

Количество пропущенных значений в каждом столбце:
Id                         2
Deal Owner Name           31
Closing Date            6950
Quality                 2255
Stage                      2
Lost Reason             5471
Page                       2
Campaign                5528
SLA                     6062
Content                 7448
Term                    9141
Source                     2
Payment Type           21099
Product                18003
Education Type         18295
Created Time               2
Course duration        18008
Months of study        20755
Initial Amount Paid    17430
Offer Total Amount     17410
Contact Name              63
City                   19084
Level of Deutsch       20344
dtype: int64


## 1. Удалите дубликаты и неактуальные столбцы.
   

Используем код, который заполняет пропущенные значения в столбцах City и Level of Deutsch значением, которое чаще всего встречается для данного Contact Name, чтобы не потерять данные о городах и уровне языка.


In [ ]:
mode_values_city = (
    deals_df
    .groupby('Contact Name')['City']
    .agg(lambda x: (
        x.mode()[0]
        if not x.mode().empty
        else None
    ))
)

deals_df['City'] = deals_df['Contact Name'].map(mode_values_city)

In [ ]:
mode_values_level = (
    deals_df
    .groupby('Contact Name')['Level of Deutsch']
    .agg(
        lambda x: (
            x.mode()[0]
            if not x.mode().empty
            else None
      ))
)

deals_df['Level of Deutsch'] = (
    deals_df['Contact Name']
    .map(mode_values_level)
)

In [ ]:
deals_df[
    deals_df.duplicated(subset=deals_df.columns[1:])
].shape

(9, 23)

In [ ]:
deals_df.drop_duplicates(
    subset=deals_df.columns[1:],
    inplace=True
)
deals_df[
    deals_df.duplicated(subset=deals_df.columns[1:])
].shape

(0, 23)

## 2. Соответствующим образом обработайте отсутствующие значения.

### Очистка столбцов

In [ ]:
deals_df.drop(
    deals_df[
        deals_df['Product'].isin(['Data Analytics', 'Find yourself in IT'])
    ].index,
    inplace=True
)

In [ ]:
deals_df.drop(
    deals_df[
        deals_df['Lost Reason'] == 'Duplicate'
    ].index,
    inplace=True
)

In [ ]:
deals_df.drop(deals_df[deals_df['Source'] == 'Test'].index, inplace=True)

In [ ]:
deals_df.drop(deals_df[deals_df['Quality'] == 'F'].index, inplace=True)

*Удаляем строки, где в колонке Product указаны значения 'Data Analytics' или 'Find yourself in IT', так как они представлены в недостаточном количестве и не несут значимой информации для анализа.*

*Удаляем строки, где в колонке Source указано значение 'Test', так как это тестовые данные, которые не относятся к реальным сделкам.
Убираем дубликаты сделок, которые были помечены как потерянные по этой причине.*

### Обработка столбца с валютой

Приведение столбцов к единому виду, очистка от символов валют

In [ ]:
print(deals_df["Offer Total Amount"].unique().tolist())
print(deals_df["Initial Amount Paid"].unique().tolist())

[nan, 2000, 9000, 11000, 3500, 4500, '€ 2.900,00', 6500, 4000, 3000, 10000, 2500, 5000, 11500, 1000, 1200, 1500, 0, '€ 11398,00', 11111]
[nan, 0, 1000, '€ 3.500,00', 500, 100, 4500, 300, 200, 2000, 11000, 4000, 3000, 3500, 11500, 1200, 1500, 5000, 600, 700, 350, 9, 400, 450]


In [ ]:
deals_df['Initial Amount Paid'] = (
    deals_df['Initial Amount Paid'].replace({9: 0})
)

*Очищаем и преобразуем данные в колонках с денежными значениями (Initial Amount Paid, Offer Total Amount), чтобы привести их к числовому формату (float) и стандартизировать валюту (евро).*

In [ ]:
def clean_currency(value):
    if isinstance(value, str):
        value = value.replace("$", "").replace("€", "").strip()
        value = value.replace("\xa0", "").replace(",", ".")
        value = value.replace(".", "", value.count(".") - 1)
        return float(value)
    elif isinstance(value, (int, float)):
        return float(value)
    return 0

deals_df["Initial Amount Paid"] = (
    deals_df["Initial Amount Paid"]
    .apply(clean_currency)
    .fillna(0)
)
deals_df["Offer Total Amount"] = (
    deals_df["Offer Total Amount"]
    .apply(clean_currency)
    .fillna(0)
)
currency_columns = ["Initial Amount Paid", "Offer Total Amount"]

for col in currency_columns:
    deals_df[col] = deals_df[col].apply(clean_currency)

deals_df.rename(
    columns={col: f"{col} (€)" for col in currency_columns},
    inplace=True
)

*Подсчитаем количество строк, где Offer Total Amount (€) меньше, чем Initial Amount Paid (€).Такие значения есть, можно сделать вывод, что данные были поменяны местами. Создаем маску для выбора этих строк. Меняем значения в этих строках местами.*

*При анализе столбцов "Offer Total Amount (€)", "Product", "Education Type" было виявлено, что если "Product" равен "UX/UI Design" и значение в столбце "Offer Total Amount (€)" равно 11000, то в соответствующих строках столбца "Education Type" будет значение "Morning".*

*Для строк, где "Offer Total Amount (€)" равен 11500, значение в столбце "Education Type" может быть только "Morning".*

In [ ]:
count = (
    deals_df["Offer Total Amount (€)"]
    < deals_df["Initial Amount Paid (€)"]
).sum()

print(
    f"Количество строк, где Offer Total Amount < Initial Amount Paid: {count}"
)

Количество строк, где Offer Total Amount < Initial Amount Paid: 59


In [ ]:
mask_amount = (
    deals_df["Offer Total Amount (€)"]
    < deals_df["Initial Amount Paid (€)"]
)
deals_df.loc[mask_amount, ["Offer Total Amount (€)", "Initial Amount Paid (€)"]] = (
    deals_df.loc[
        mask_amount,
        ["Initial Amount Paid (€)", "Offer Total Amount (€)"]
    ].values
)
count = (
    deals_df["Offer Total Amount (€)"]
    < deals_df["Initial Amount Paid (€)"]
).sum()

print(f"Значения в {mask_amount.sum()} строках успешно поменялись местами.")
print(
    f"Количество строк, где Offer Total Amount < Initial Amount Paid: {count}"
)

Значения в 59 строках успешно поменялись местами.
Количество строк, где Offer Total Amount < Initial Amount Paid: 0


In [ ]:
mask_education = (
    (deals_df["Offer Total Amount (€)"] == 11000) &
    (deals_df["Product"] == "UX/UI Design") &
    (
        deals_df["Education Type"].isin(["", "Evening"]) |
        deals_df["Education Type"].isna()
    )
)
deals_df.loc[mask_education, "Education Type"] = "Morning"

In [ ]:
mask_education_2 = (
    (deals_df["Offer Total Amount (€)"] == 11500) &
    (deals_df["Education Type"].isna())
)
deals_df.loc[mask_education_2, "Education Type"] = "Morning"

### Удаление полностью пустых строк

In [ ]:
deals_df.tail(2)

Id Deal Owner Name Closing Date Quality Stage Lost Reason Page  \
21593  NaN             NaN          NaN     NaN   NaN         NaN  NaN   
21594  NaN             NaN          NaN     NaN   NaN         NaN  NaN   

      Campaign  SLA Content Term Source Payment Type Product Education Type  \
21593      NaN  NaN     NaN  NaN    NaN          NaN     NaN            NaN   
21594      NaN  NaN     NaN  NaN    NaN          NaN     NaN          #REF!   

      Created Time  Course duration  Months of study  Initial Amount Paid (€)  \
21593          NaN              NaN              NaN                      0.0   
21594          NaN              NaN              NaN                      0.0   

       Offer Total Amount (€) Contact Name City Level of Deutsch  
21593                     0.0          NaN  NaN              NaN  
21594                     0.0          NaN  NaN              NaN

In [ ]:
deals_df.replace(r"^\s*$", np.nan, regex=True, inplace=True)
deals_df.replace("#REF!", np.nan, inplace=True)

deals_df = deals_df[~((deals_df.isna()) | (deals_df == 0.0)).all(axis=1)]
deals_df.reset_index(drop=True, inplace=True)

print(deals_df.shape)

(19657, 23)


### Переводим в секунды значения столбца SLA

In [ ]:
deals_df['SLA'].value_counts()

SLA
00:11:09            6
00:10:46            6
00:10:11            6
00:12:52            5
00:17:43            5
                   ..
01:23:14            1
01:14:13            1
01:41:52            1
05:44:49            1
4 days, 22:47:14    1
Name: count, Length: 12830, dtype: int64

In [ ]:
def convert_to_seconds(x):
    if pd.isna(x):
        return np.nan
    elif isinstance(x, datetime.time):
        return x.hour * 3600 + x.minute * 60 + x.second
    elif isinstance(x, datetime.timedelta):
        return x.total_seconds()

deals_df['SLA Seconds'] = deals_df["SLA"].apply(convert_to_seconds)
deals_df['SLA Seconds'].value_counts()

SLA Seconds
669.0       6
646.0       6
611.0       6
772.0       5
1063.0      5
           ..
4994.0      1
4453.0      1
6112.0      1
20689.0     1
427634.0    1
Name: count, Length: 12830, dtype: int64

In [ ]:
deals_df.drop(columns=['SLA'], inplace=True)

### Обработка столбца City
Удаление символов "-"

In [ ]:
unique_cities = deals_df["City"].unique()
print("Уникальные значения в столбце 'City':")
print(unique_cities)

total_values = deals_df["City"].count()
print("Сумма всех значений (без NaN):", total_values)

dash_count = (deals_df["City"] == "-").sum()
print("Количество значений '-' в столбце 'City':", dash_count)

Уникальные значения в столбце 'City':
[None 'Berlin' 'Lahnstein' 'Crailsheim' 'Prenzlau' 'Dortmund' 'Stuttgart'
 'München' 'Wien' 'Offenbach am Main' 'Eberbach' 'Görlitz' 'Pfedelbach'
 'Unterhaching' 'Karl-Liebknecht str. 24, Hildburghausen, Thüringen'
 'Sonneberg' 'Düsseldorf' '-' 'Pirmasens' 'Lutherstadt Wittenberg'
 'Gommern' 'Rüdesheim am Rhein' 'Dresden' 'Passau' 'Leipzig' 'Bielefeld'
 'Karlsruhe' 'Hamm' 'Wermelskirchen' 'Waldbronn' 'Kassel' 'Wenzenbach'
 'Duisburg' 'Witzeeze' 'Leinfelden-Echterdingen' 'Kerpen' 'Ahrensburg'
 'Schwandorf' 'Merseburg' 'Dabrowa Gornicza' 'Pommelsbrunn' 'Regensburg'
 'Herzogenrath' 'Mainz' 'Podskalie' 'Zinnowitz' 'Quedlinburg'
 'Poland , Gdansk , Al. Grunwaldzka 7, ap. 1a' 'Wolfsburg' 'Bochum'
 'Weilburg' 'Weingarten' 'Wiesbaden' 'Dillenburg' 'Neu-Ulm' 'Celle'
 'Neu Wulmstorf' 'Marburg' 'Lauter-Bernsbach' 'Bonn' 'Riedstadt'
 'Rosenheim' 'Herford' 'Mönchengladbach' 'Neuburg' 'Rostock'
 'Bad Oeynhausen' 'Chemnitz' 'Diez' 'Leutershausen' 'Nürnberg'
 'Kor

In [ ]:
deals_df["City"] = deals_df["City"].replace("-", np.nan)
total_values = deals_df["City"].count()
print("Сумма всех значений (без NaN):", total_values)

dash_count = (deals_df["City"] == "-").sum()
print("Количество значений '-' в столбце 'City':", dash_count)

Сумма всех значений (без NaN): 2675
Количество значений '-' в столбце 'City': 0


### Обработка столбца с уровнями языков

In [ ]:
deals_df['Level of Deutsch'].value_counts()

Level of Deutsch
B1                                                      269
б1                                                      135
в1                                                      124
b1                                                      121
Б1                                                      118
                                                       ... 
А2(ждет итоги Б1)                                         1
в1-ня , в1-ая                                             1
ня-0, но англ B2+                                         1
В                                                         1
точно уровень не знаю, но говорить могу - учила сама      1
Name: count, Length: 214, dtype: int64

In [ ]:
unique_values_1 = deals_df["Level of Deutsch"].unique().tolist()
display(unique_values_1)

[None,
 'б1',
 'в1',
 'A2',
 'в2',
 'b1',
 'В1',
 'B1',
 'в1-в2',
 'А2-В1',
 'А2 ( Б1 в июне)',
 '-',
 'B2',
 'C2',
 'с1',
 'Б1',
 'Б2',
 'b1 ждет результаты',
 'а2',
 'Б1( может будет)',
 'а1',
 'С1',
 'a2 (b1 экзамен 15 июня)',
 'а0',
 'б2',
 'А2',
 'B1 будет в феврале 2025',
 'Detmold, Paulinenstraße 95, 32756',
 'Сам оценивает на B2, 13 лет живет в Германии',
 'В1-В2',
 'Б1 ( ждет Б2)',
 'lэкзамен - 6 июля на В1. курсы вечером (но уверенно говорит на B1)',
 'Гражданка Германии уже год в Германии Учит немецкий и в сентябре b1 через гос-во проходит, а не через ДЖЦ, вечером учится 3 р в неделю с 18 до 21',
 'B1 в процессе обучения',
 'ЯЗ: нем В1 был экз 03.05 повтор и сейчас ждет результаты. Технический англ был. А1 сейчас. ОБР: 2 во информационные и комп сети - инженер системоте',
 'В1 в сентябре',
 'Нет',
 0,
 'Ждем B1',
 'А1 сертиф, но по факту А2',
 'a2',
 'Пока А2, сдает 17 05 B1',
 'окончание 13.06 курса на b1',
 'A1',
 'b2',
 'В2',
 'Thorn-Prikker-Str. 30, Hagen, 58093',
 'нуле

In [ ]:
replacements = {
    None: "Unknown",
    "-": "Unknown",
    "Нет": "Unknown",
    0: "Unknown",
    "nan": "Unknown",
    ".": "Unknown",
    "?": "Unknown",
    "УТОЧНИТЬ!": "Unknown",
    "УТОЧНИТЬ": "Unknown",
    "никакой": "Unknown",
    "Гражданка": "B1",
    "гражданка": "B1",
    "90": "Unknown",
    "31.05.2024": "Unknown",
    "f2": "Unknown",
    "Бй": "B1",
    "Б10Б2": "B1",
    "В": "B1",

    "Detmold, Paulinenstraße 95, 32756": "Unknown",
    "Thorn-Prikker-Str. 30, Hagen, 58093": "Unknown",
    "Paderborn 33102, Schwabenweg 10": "Unknown",
    "Lichtenfelser Straße 25, Untersiemau 96253": "Unknown",

    **{variant: level for level in ["A0", "A1", "A2", "B1", "B2", "C1", "C2"]
        for variant in [
            level.lower(),
            level.upper(),
            level.lower().replace('a', 'а').replace('b', 'б').replace('c', 'с'),
            level.upper().replace('A', 'А').replace('B', 'Б').replace('C', 'С')
        ]},

    "а1-а2": "A1",
    "a0-a1": "A0",
    "А2-В1": "A2",
    "в1-в2": "B1",
    "в2-с1": "B2",
    "b2-c1": "B2",
    "b2-c2": "B2",
    "b1+": "B1",
    "б2+": "B2",
    "B2+": "B2",
    "B1+": "B1",
    "А2+": "A2",

    **{key: "B1" for key in [
        "b1 ждет результаты", "b1 ждет экзамен", "b1 ждет серт", "B1 (ждет.*)",
        "b1 (ждет.*)", "Ждем B1", "Ждет B1", "B1 в процессе", "B1 будет.*",
        "b1 экзамен.*", "B1 должен.*", "B1 еще нет.*", "Учиться на B1",
        "b1 (b2.*)", "B1 (B2.*)", "Б1(.*Б2)", "В1(.*В2)"
    ]},

    **{key: "A2" for key in [
        "А2 (.*Б1)", "А2 (.*В1)", "a2 (.*b1)", "А2, в процессе Б1",
        "А2(.*Б1)", "А2(.*В1)", "А2 скоро.*", "А2 заканчив.*"
    ]},

    "Сам оценивает на B2, 13 лет живет в Германии": "B2",
    "25 лет живет в Германии": "B1",
    "нулевой уровень, только пошел на курсы.": "A0",
    "разговорный из украины, без сертификата": "A2",
    "lэкзамен - 6 июля на В1. курсы вечером (но уверенно говорит на B1)": "B1",
    "С1 -ая , Ня -а1": "C1",
    "ня а2, ая в1": "B1",
    "ня-0, но англ B2+": "A0",
    "а2, англ B1": "A2",
    "НЯ - В1, АЯ - В1": "B1",
    "немецкий - а1-а2, англ b1-b2": "A2",
    "А2 нем -В2 англ": "A2",
    "B1 немецкий и английский Advance": "B1",
    "не учил": "Unknown",
    "не учила ( разговорный) сразу пошла работать": "A2",
    "B1 (почти, не сдала чуть) + англ В1": "B1",
    "B2 - не сдал": "B1",
    "B2 (не сдал)": "B1",

    **{key: "B1" for key in [
        "B1 будет в феврале 2025", "b1 ждет результаты",
        "b1 (b2 ждет серт)", "Б1( может будет)",
        "Б1 ( ждет Б2)", "B1 будет в феврале 2025",
        "B1 в сентябре", "b1 ждет сертификат",
        "B1 (B2 должна до конца февраля получить)"
    ]},

    **{key: "B2" for key in [
        "5 июля 2024 сдает экз на В2",
        "В январе - В2 сдает", "b2 (с1 экзамен 16 февраля)"
    ]},
}

deals_df['Level of Deutsch'] = deals_df['Level of Deutsch'].map(replacements)
deals_df['Level of Deutsch'].fillna('Unknown', inplace=True)

In [ ]:
unique_values = deals_df["Level of Deutsch"].unique().tolist()
unique_values

['Unknown', 'B1', 'A2', 'B2', 'C2', 'C1', 'A1', 'A0']

In [ ]:
deals_df['Level of Deutsch'].value_counts()

Level of Deutsch
Unknown    18580
B1           703
A2           157
B2           149
C1            33
A1            22
A0             9
C2             4
Name: count, dtype: int64

## 3. Преобразуйте типы данных для таких столбцов, как даты и числовые значения.

In [ ]:
deals_df['Closing Date'] = pd.to_datetime(
    deals_df['Closing Date'],
    format='%d.%m.%Y',
    errors='raise'
)
deals_df['Created Time'] = pd.to_datetime(
    deals_df['Created Time'],
    format='%d.%m.%Y %H:%M',
    errors='raise'
)

In [ ]:
display(deals_df.info(), deals_df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19657 entries, 0 to 19656
Data columns (total 23 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Id                       19657 non-null  object        
 1   Deal Owner Name          19628 non-null  object        
 2   Closing Date             13057 non-null  datetime64[ns]
 3   Quality                  17426 non-null  object        
 4   Stage                    19657 non-null  object        
 5   Lost Reason              14267 non-null  object        
 6   Page                     19657 non-null  object        
 7   Campaign                 15462 non-null  object        
 8   Content                  13640 non-null  object        
 9   Term                     11944 non-null  object        
 10  Source                   19657 non-null  object        
 11  Payment Type             479 non-null    object        
 12  Product                  3513 no

None

Id                             0
Deal Owner Name               29
Closing Date                6600
Quality                     2231
Stage                          0
Lost Reason                 5390
Page                           0
Campaign                    4195
Content                     6017
Term                        7713
Source                         0
Payment Type               19178
Product                    16144
Education Type             16305
Created Time                   0
Course duration            16144
Months of study            18822
Initial Amount Paid (€)        0
Offer Total Amount (€)         0
Contact Name                  47
City                       16982
Level of Deutsch               0
SLA Seconds                 4800
dtype: int64

 Сохранение преобразованного датасета

In [ ]:
deals_df.to_excel('/content/Processed_Deals_df.xlsx', index=False)

# **Spend**


In [ ]:
print("Первые 5 строк:")
display(spend_df.head())

print("Информация о датасете:")
display(spend_df.info())

print("Описательная статистика :")
display(spend_df.describe())

print("Количество пропущенных значений в каждом столбце:")
print(spend_df.isnull().sum())

Первые 5 строк:


Date        Source               Campaign  Impressions  Spend  Clicks  \
0 2023-07-03    Google Ads         gen_analyst_DE            6   0.00       0   
1 2023-07-03    Google Ads  performancemax_eng_DE            4   0.01       1   
2 2023-07-03  Facebook Ads                    NaN            0   0.00       0   
3 2023-07-03    Google Ads                    NaN            0   0.00       0   
4 2023-07-03           CRM                    NaN            0   0.00       0   

  AdGroup   Ad  
0     NaN  NaN  
1     NaN  NaN  
2     NaN  NaN  
3     NaN  NaN  
4     NaN  NaN

Информация о датасете:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20779 entries, 0 to 20778
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         20779 non-null  datetime64[ns]
 1   Source       20779 non-null  object        
 2   Campaign     14785 non-null  object        
 3   Impressions  20779 non-null  int64         
 4   Spend        20779 non-null  float64       
 5   Clicks       20779 non-null  int64         
 6   AdGroup      13951 non-null  object        
 7   Ad           13951 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 1.3+ MB


None

Описательная статистика :


Date    Impressions         Spend  \
count                          20779   20779.000000  20779.000000   
mean   2024-01-14 22:32:40.864334080    2458.203475      7.195892   
min              2023-07-03 00:00:00       0.000000      0.000000   
25%              2023-10-13 00:00:00       0.000000      0.000000   
50%              2024-01-27 00:00:00      63.000000      0.580000   
75%              2024-04-16 00:00:00     709.000000      5.750000   
max              2024-06-21 00:00:00  431445.000000    774.000000   
std                              NaN   11442.528075     26.760080   

             Clicks  
count  20779.000000  
mean      23.990616  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%       12.000000  
max     2415.000000  
std       85.245714

Количество пропущенных значений в каждом столбце:
Date              0
Source            0
Campaign       5994
Impressions       0
Spend             0
Clicks            0
AdGroup        6828
Ad             6828
dtype: int64


Соответствующим образом обработайте отсутствующие значения.

In [ ]:
spend_df = spend_df[~((spend_df["Impressions"] == 0) &
                      (spend_df["Spend"] == 0) &
                      (spend_df["Clicks"] == 0))]

*Удаляем строки, где одновременно равны нулю три колонки: Impressions, Spend и Clicks, которые не несут полезной информации для анализа.*

 Удалите дубликаты и неактуальные столбцы.



In [ ]:
spend_df[spend_df.duplicated(subset=spend_df.columns[1:])].shape

(1440, 8)

In [ ]:
spend_df.drop_duplicates(subset=spend_df.columns[1:], inplace=True)

Преобразуйте типы данных для таких столбцов, как даты и числовые
значения.


In [ ]:
spend_df["Date"] = pd.to_datetime(spend_df["Date"], errors="raise")

In [ ]:
display(spend_df.info(), spend_df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
Index: 14967 entries, 0 to 20778
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         14967 non-null  datetime64[ns]
 1   Source       14967 non-null  object        
 2   Campaign     13728 non-null  object        
 3   Impressions  14967 non-null  int64         
 4   Spend        14967 non-null  float64       
 5   Clicks       14967 non-null  int64         
 6   AdGroup      12894 non-null  object        
 7   Ad           12894 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 1.0+ MB


None

Date              0
Source            0
Campaign       1239
Impressions       0
Spend             0
Clicks            0
AdGroup        2073
Ad             2073
dtype: int64

In [ ]:
spend_df.to_excel('/content/Processed_Spend_df.xlsx', index=False)

# **Описательная статистика**

In [ ]:
processed_contacts_df = pd.read_excel(
    'Processed_Contacts_df.xlsx',
    dtype={'Id': str}
)
processed_calls_df = pd.read_excel(
    'Processed_Calls_df.xlsx',
    dtype={'Id': str, "CONTACTID": str}
)
processed_spend_df = pd.read_excel('Processed_Spend_df.xlsx')
processed_deals_df = pd.read_excel(
    'Processed_Deals_df.xlsx',
    dtype={'Id': str, 'Contact Name': str}
)

## 1. Рассчитайте сводную статистику
 (среднее значение, медиана, режим, диапазон) для числовых полей.


In [ ]:
def describe_numeric_data(df_dict):
    for df_name, df in df_dict.items():
        numeric_df = df.select_dtypes(include=['number'])

        if numeric_df.empty:
            print(f"⚠️ В таблице {df_name} нет числовых данных.")
            continue

        stats = pd.DataFrame({
            "Среднее": numeric_df.mean(),
            "Медиана": numeric_df.median(),
            "Режим": numeric_df.mode().iloc[0] if not numeric_df.mode().empty else None,
            "Мин": numeric_df.min(),
            "Макс": numeric_df.max(),
            "Диапазон": numeric_df.max() - numeric_df.min()
        }).round(2)

        print(f"📊 Сводная статистика для {df_name}:")
        print(stats.to_string())
        print("\n" + "=" * 80 + "\n")

tables = {
    "processed_deals_df": processed_deals_df,
    "processed_spend_df": processed_spend_df,
    "processed_calls_df": processed_calls_df,
    "processed_contacts_df": processed_contacts_df
}

describe_numeric_data(tables)

📊 Сводная статистика для processed_deals_df:
                           Среднее  Медиана  Режим  Мин        Макс    Диапазон
Course duration              10.20     11.0   11.0  6.0        11.0         5.0
Months of study               5.45      5.0    6.0  0.0        11.0        11.0
Initial Amount Paid (€)     191.36      0.0    0.0  0.0     11000.0     11000.0
Offer Total Amount (€)     1490.06      0.0    0.0  0.0     11500.0     11500.0
SLA Seconds              111722.39  19603.0  611.0  3.0  26908464.0  26908461.0


📊 Сводная статистика для processed_spend_df:
             Среднее  Медиана  Режим  Мин      Макс  Диапазон
Impressions  3411.95   286.00    0.0  0.0  431445.0  431445.0
Spend           9.97     2.35    0.0  0.0     774.0     774.0
Clicks         30.98     4.00    0.0  0.0    2415.0    2415.0


📊 Сводная статистика для processed_calls_df:
                            Среднее  Медиана  Режим  Мин    Макс  Диапазон
Call Duration (in seconds)   170.33      9.0    0.0  0.0  

***Выводы:***
* processed_deals_df

*Course duration имеет среднее значение 10.20, медиана 11.0, мода тоже 11.0. Это говорит о том, что большинство курсов длятся около 11 месяцев, но есть и более короткие, минимум 6 месяцев.*
*Months of study: имеет минимум — 0. Возможно, это ошибка данных или какие-то курсы не требуют изучения.* *Максимум 11, это может указывать на то, что студенты учатся от нуля до 11 месяцев, но большинство около 5-6 месяцев. *
*SLA Seconds: минимальное значение 3 секунды, максимум 26 миллионов. Это может означать, что есть как очень быстрые, так и крайне медленные обработки заявок. *
Initial Amount Paid (€): *большинство* клиентов не вносили предоплату, но есть крупные платежи (максимум 11,500 €).
*  processed_spend_df

*Impressions, Spend, Clicks. Везде медианы и моды значительно ниже среднего, что указывает на правостороннее смещение. Большинство значений низкие, но есть редкие случаи с очень высокими показателями.Это может быть связано с рекламными кампаниями, где большинство объявлений получают мало взаимодействий, но некоторые становятся вирусными.*
* processed_calls_df

*Call Duration (in seconds) — большинство звонков очень короткие, но есть несколько очень длинных, до 7625 секунд (более 2 часов). Возможно, это ошибки в данных или реальные длинные консультации. Scheduled in CRM: все значения 0 или 1, среднее 0.00 — значит, почти все звонки не запланированы в CRM.*



## 2. Анализируйте категориальные поля,
 такие как качество, стадия, источник и
продукт.


In [ ]:
def analyze_categorical_columns(df, columns):
    for col in columns:
        if col in df.columns:
            print(f"{col} (Уникальных значений: {df[col].nunique()})")
            print("" + "-" * 33 + "\n")
            abs_counts = df[col].value_counts()
            total = abs_counts.sum()
            for category, count in abs_counts.items():
                percentage = (count / total) * 100 if total != 0 else 0.0
                print(f"  - {category}: \033[1;30m{percentage:.2f}% ({count})\033[0m")
        else:
            print(f"\n\033[1;31m{col} - Колонка отсутствует в DataFrame\033[0m")
        print("\n" + "=" * 80 + "\n")

categorical_columns = ['Quality', 'Stage', 'Source', 'Product']
analyze_categorical_columns(processed_deals_df, categorical_columns)

Quality (Уникальных значений: 5)
---------------------------------

  - E - Non Qualified: 35.06% (6110)
  - D - Non Target: 34.29% (5975)
  - C - Low: 19.45% (3389)
  - B - Medium: 8.81% (1536)
  - A - High: 2.39% (416)


Stage (Уникальных значений: 13)
---------------------------------

  - Lost: 70.79% (13916)
  - Call Delayed: 11.04% (2170)
  - Registered on Webinar: 10.51% (2066)
  - Payment Done: 4.33% (852)
  - Waiting For Payment: 1.64% (323)
  - Qualificated: 0.65% (128)
  - Registered on Offline Day: 0.43% (85)
  - Need to Call - Sales: 0.16% (32)
  - Need To Call: 0.16% (31)
  - Test Sent: 0.13% (25)
  - Need a consultation: 0.12% (23)
  - New Lead: 0.03% (5)
  - Free Education: 0.01% (1)


Source (Уникальных значений: 12)
---------------------------------

  - Facebook Ads: 24.06% (4729)
  - Google Ads: 20.93% (4114)
  - Tiktok Ads: 10.19% (2003)
  - SMM: 8.49% (1669)
  - Youtube Ads: 8.23% (1618)
  - Organic: 7.61% (1496)
  - CRM: 7.40% (1455)
  - Bloggers: 5.46% (1074)
  

***Выводы:***
* Качество лидов (Quality):

*E (Non Qualified) и D (Non Target) составляют 69.35%  — большинство лидов не соответствуют целевой аудитории или не прошли квалификацию.Только 2.39% лидов имеют высокое качество (A - High), что крайне мало.*

*Рекомендации:*
*Пересмотреть критерии таргетинга рекламы.Улучшить процесс квалификации лидов.Увеличить долю лидов категорий B и A.*
* Стадии сделок (Stage):

*70.79% сделок потеряны (Lost) — критически высокий показатель.Только 4.33% доходят до оплаты (Payment Done), а 1.64% ожидают оплаты (Waiting For Payment).Стадии, связанные с активными действиями (Need to Call, Qualificated), составляют менее 1%.*

*Рекомендации:*
*Анализ причин потерь на стадии Lost.Усилить работу с лидами на стадиях Qualificated и Need to Call.Оптимизировать процесс вебинаров (Registered on Webinar — 10.51%).*

* Источники лидов (Source):

*Онлайн-реклама доминирует: Facebook Ads (24.06%), Google Ads (20.93%), TikTok Ads (10.19%), Органические источники (Organic, Bloggers, Telegram) занимают 18.12%, Offline-источники почти отсутствуют (0.01%).*

*Рекомендация:*
*Сравнить ROI по источникам, стоит развивать партнёрские программы (Partnership — 1.02%) и оффлайн-активности.*
* Продукты (Product):

*Digital Marketing — ключевое направление.*

*Рекомендации:*
*Усилить продвижение UX/UI Design и Web Developer.*
